src/clusterfuzz/_internal/system/archive.py

```diff
@@ -163,171 +370,46 @@ def has_extension(extensions):
   return ArchiveType.UNKNOWN
 
 
-def get_file_list(archive_path, archive_obj=None, file_match_callback=None):
-  """List all files in an archive."""
-  return [
-      f.name for f in iterator(
-          archive_path, archive_obj, file_match_callback, should_extract=False)
-  ]
-
-
-def get_first_file_matching(search_string, archive_obj, archive_path):
-  """Returns the first file with a name matching search string in archive."""
-  for current_file in get_file_list(archive_path, archive_obj):
-    # Exclude MAC resouce forks.
-    if current_file.startswith('__MACOSX/'):
-      continue
-
-    if search_string in current_file:
-      return current_file
-
-  return None
-
-
 def is_archive(filename):
   """Return true if the file is an archive."""
   return get_archive_type(filename) != ArchiveType.UNKNOWN
 
 
-def unpack(archive_path,
-           output_directory,
-           trusted=False,
-           file_match_callback=None):
-  """Extracts an archive into the target directory."""
-  if not os.path.exists(archive_path):
-    logs.log_error('Archive %s not found.' % archive_path)
-    return False
-
-  # If the output directory is a symlink, get its actual path since we will be
-  # doing directory traversal checks later when unpacking the archive.
-  output_directory = os.path.realpath(output_directory)
-
-  archive_filename = os.path.basename(archive_path)
-  error_occurred = False
-
-  # Choose to unpack all files or ones matching a particular regex.
-  file_list = get_file_list(
-      archive_path, file_match_callback=file_match_callback)
+def unpack(reader, output_dir, trusted=False, file_match_callback=None):
```

Comment:
This seems like it could be a non-virtual method on `ArchiveReader`?

If you're strongly against adding concrete behavior to the base class, then consider the following design:

```py
class ArchiveReader:
  def __init__(impl: _ArchiveReaderImpl):
    self._impl = impl

  def extract(...):
    if traversal:
      return None
    return self._impl.extract(...)

  def unpack(...):
    ... do stuff ...
    for file in list:
      self._impl.extract(...)
    ...

class _ArchiveReaderImpl(abc.ABC):
  @abc.abstractmethod
  def extract(...):
    raise NotImplementedError

class _TarArchiveReaderImpl(ArchiveReaderImpl):
  def extract(...):
    ...

def open(file):
  if is_tar:
    return ArchiveReader(_TarArchiveReaderImpl(file))
  ...
```

Note that this lends itself to testing the shared code in `ArchiveReader` by injecting a mock impl.

Reply:
I agree with you, but if you're ok with it, I'd like to do this in a second time, mostly because I'll have to design things in a way that it can be extended to take `.runtime_deps` into account.

Reply:
Fair enough!